File: sources/CPP/include/Ers/Visualization/Texture.cpp
#include "Texture.h"

#include "Ers/Api.h"

namespace Ers
{
    Texture::Texture()
    {
        corePtr = Ers::Engine::ERS_Texture_Create();
    }

    Texture::Texture(const std::string& path)
    {
        corePtr = Ers::Engine::ERS_Texture_Create();
        Ers::Engine::ERS_Texture_Load(corePtr, path.c_str());
    }

    void Texture::Destroy()
    {
        Ers::Engine::ERS_Texture_Release(CorePtr());
    }

    void* Texture::CorePtr()
    {
        return corePtr;
    }

    const void* const Texture::CorePtr() const
    {
        return corePtr;
    }
} // namespace Ers

File: sources/CSharp/src/Ers/Visualization/Texture.cs
using Ers.Engine;

namespace Ers
{
    /// <summary>
    /// A texture on the GPU.
    /// </summary>
    public class Texture : IDisposable
    {
        /// <summary>
        /// Native pointer to the core instance.
        /// </summary>
        public IntPtr CorePtr;

        /// <summary>
        /// Create an empty texture.
        /// </summary>
        public Texture() { CorePtr = ErsEngine.ERS_Texture_Create(); }

        /// <summary>
        /// Create a texture from an image file.
        /// </summary>
        /// <param name="path">The path to the image file.</param>
        public Texture(string path)
        {
            CorePtr      = ErsEngine.ERS_Texture_Create();
            var pathUtf8 = path.ToUtf8NullTerminated();
            unsafe
            {
                fixed(byte* pathByte = pathUtf8)
                {
                    ErsEngine.ERS_Texture_Load(CorePtr, pathByte);
                }
            }
        }

        internal Texture(IntPtr corePtr) { CorePtr = corePtr; }

        /// <summary>
        /// Finalizer.
        /// </summary>
        ~Texture() => DisposeInner();

        /// <summary>
        /// Indicate the Texture is no longer required and can be freed from memory.
        /// </summary>
        public void Dispose()
        {
            DisposeInner();
            GC.SuppressFinalize(this);
        }

        private void DisposeInner()
        {
            if (CorePtr != IntPtr.Zero)
            {
                ErsEngine.ERS_Texture_Release(CorePtr);
                CorePtr = IntPtr.Zero;
            }
        }
    }
}
